lib/assessment.js

```diff
@@ -81,19 +84,40 @@ module.exports = {
                         const question_id = null; // use instance_question_id to determine the question
                         const options = {};
                         const require_open = true;
-                        question._ensureVariantWithClient(client, question_id, instance_question_id, user_id, authn_user_id, course, options, require_open, (err, _variant) => {
+                        question._ensureVariantWithClient(client, question_id, instance_question_id, user_id, authn_user_id, course, options, require_open, (err, variant) => {
                             if (ERR(err, callback)) return;
+                            if (variant.broken) {
+                                brokenVariants.push(variant);
+                            }
                             callback(null);
                         });
                     }, (err) => {
                         if (ERR(err, callback)) return;
                         callback(null);
                     });
                 },
+                (callback) => {
+                    if (brokenVariants.length > 0) {
+                        // At least one variant is broken, so delete the assessment instance
+                        var params = [
+                            assessment_instance_id,
+                            authn_user_id,
+                        ];
+                        sqldb.callWithClientOneRow(client, 'assessment_instances_delete', params, (err, _result) => {
+                            if (ERR(err, callback)) return;
+                        });
+                    }
+                    callback(null);
```

Comment:
This `callback(null)` call needs to go inside the `assessment_instances_delete` callback so that we only return and commit the transaction after the delete has occurred. I'm fixing this.

Reply:
Thanks. So it's ok not to have any `callback(null)` at all in the case of no broken variants?

Reply:
Good point. We actually need something like this:
```
if (brokenVariants.length > 0) {
    sqldb.call(...) {
        callback(null);
    }
} else {
    callback(null);
}
```

This way if we have broken variants, we first delete the assessment_instance and then call `callback` to continue. If we don't have broken variants we immediately continue.

You are correct that we always need to call `callback()`. But we shouldn't call it until we've finished with all the work we are going to do, if that makes sense?